sw/device/silicon_creator/lib/manifest.h

```diff
@@ -192,7 +197,7 @@ typedef struct manifest {
    * Unix timestamp that gives the creation time of the image, seconds since
    * 00:00:00 on January 1, 1970 UTC (the Unix Epoch).
    */
-  uint64_t timestamp;
+  uint32_t timestamp[2];
```

Comment:
Why can't we represent this as `uint64_t`?

Should we even have a timestamp field?  Depending on how this is set, it can lead to a non-reproducible build (ie: if the value is set with something like `Instant::now()` or `datetime.utcnow()`).  To be reproducible, this value is going to have to be set manually in the manifest, probably to something like the datetime of the git commit/tag that we're building.

Reply:
having a `uint64_t` pushes the alignment of this struct to 64-bits, requiring a 32-bit padding in some cases. 2 x 32-bit -> 64-bit is a trivial operation. I think it's a nice to have piece of information with a negligible amount of overhead, esp. when compared to the modulus we store for OOB signatur verification.

Reply:
The struct itself is 64-bit aligned in its use-case within OpenTitan.  Can we not choose a location within the manifest that satisfies the alignment requirement of the `timestamp` field such that there is no internal padding?

I'm having a really hard time accepting a decomposed representation for `uint64_t` in ${current_year}.

Reply:
I support making the timestamp a function of the commit. If our builds are reproducible, this should have roughly the same utility as a true "now()" timestamp.

As for alignment, I noticed we can use `uint64_t` and recover the struct's 32-bit alignment by packing *only the timestamp field*. I believe the concern with using `packed` is its implication on alignment, but empirically, I think it's actually doing what we want here.

Alternatively, a practical solution is to make the field `char[8]` and define an accessor function `uint64_t manifest_get_timestamp(const manifest_t*);`. 

```diff
diff --git a/sw/device/silicon_creator/lib/manifest.h b/sw/device/silicon_creator/lib/manifest.h
index 573df5e10..f7fe8982c 100644
--- a/sw/device/silicon_creator/lib/manifest.h
+++ b/sw/device/silicon_creator/lib/manifest.h
@@ -5,6 +5,7 @@
 #ifndef OPENTITAN_SW_DEVICE_SILICON_CREATOR_LIB_MANIFEST_H_
 #define OPENTITAN_SW_DEVICE_SILICON_CREATOR_LIB_MANIFEST_H_
 
+#include <stdalign.h>
 #include <stddef.h>
 
 #include "sw/device/lib/base/macros.h"
@@ -197,7 +198,8 @@ typedef struct manifest {
    * Unix timestamp that gives the creation time of the image, seconds since
    * 00:00:00 on January 1, 1970 UTC (the Unix Epoch).
    */
-  uint32_t timestamp[2];
+  uint64_t timestamp __attribute__ ((__packed__));
+  //char timestamp[8];
   /**
    * Binding value used by key manager to derive secret values.
    *
@@ -246,6 +248,7 @@ OT_ASSERT_MEMBER_OFFSET(manifest_t, code_start, 8776);
 OT_ASSERT_MEMBER_OFFSET(manifest_t, code_end, 8780);
 OT_ASSERT_MEMBER_OFFSET(manifest_t, entry_point, 8784);
 OT_ASSERT_SIZE(manifest_t, CHIP_MANIFEST_SIZE);
+static_assert(alignof(manifest_t) == sizeof(uint32_t), "");
```

Reply:
> The struct itself is 64-bit aligned in its use-case within OpenTitan. 

I don't think we are talking about the same thing. I'm specifically referring to the alignment requirement of the struct. 

> Can we not choose a location within the manifest that satisfies the alignment requirement of the `timestamp` field such that there is no internal padding?

We can, I've hit this I think the 3rd time. any time I introduce a new field before the special location or move things around I have to adjust the offsets as well as opentitantool which is unnecessary toil. 

> I'm having a really hard time accepting a decomposed representation for `uint64_t` in ${current_year}.

That's OK. We have bigints that are just  word arrays.

Reply:
> I support making the timestamp a function of the commit. If our builds are reproducible, this should have roughly the same utility as a true "now()" timestamp.

I know you're responding to @cfrantz's comment but this is orthogonal to this PR. Every field in the manifest must be fixed for the build to be reproducible.

> +  uint64_t timestamp __attribute__ ((__packed__));

`__packed__` is no bueno (not sure if bindgen supports this).

We can just use `uint32_t[2]`s without increasing the alignment requirement of the struct. The use case of this field is helping tooling and maybe reporting post-boot and if anyone needs to get the 64-bit value:
```
((uint64_t)ts[1] << 32) | ts[0]
```

Reply:
> > The struct itself is 64-bit aligned in its use-case within OpenTitan.
> 
> I don't think we are talking about the same thing. I'm specifically referring to the alignment requirement of the struct.

That's what I'm referring to: the struct is always at least 64-bit aligned when placed in OT's flash.  In fact, all instances of the manifest in the flash are going to be _at least_ 2K aligned because that is the flash sector size.  I don't see making the struct have a 64-bit alignment requirement as far as C or Rust is concerned to be a problem.

> We can, I've hit this I think the 3rd time. any time I introduce a new field before the special location or move things around I have to adjust the offsets as well as opentitantool which is unnecessary toil.

We're not _supposed_ to be modifying the manifest - we modified the front end of the struct out of expediency to get the SPX implementation done with the intention of reverting/correcting that change.  Apart from appending a fields onto the end of the struct, changing the struct is supposed to be toilsome.

> > I'm having a really hard time accepting a decomposed representation for `uint64_t` in ${current_year}.
> 
> That's OK. We have bigints that are just word arrays.

False equivalence.  Bigints are composites and, to my knowledge, have never had a built-in language type in any systems programming language.  The 64-bit integer has been a standard part of the C language since C99.

Back in the olden times, C compilers gained 64-bit integer support.  A certain software house who makes a popular PC operating system belligerently refused to update their compiler even though 64-bit ints were standardized and they had APIs to return 64-bit quantities for things like high-resolution timers.  Writing a portable library to deal with high-res timers across different OSes and compilers was a tedious (but not terribly difficult) task.  Even today, one has to do some hoop-jumping to deal with this on that platform (see [here](https://learn.microsoft.com/en-us/windows/win32/api/profileapi/nf-profileapi-queryperformancecounter), [here](https://learn.microsoft.com/en-us/windows/win32/api/winnt/ns-winnt-large_integer-r1) and [here](https://learn.microsoft.com/en-us/windows/win32/winprog/windows-data-types)).

We literally going to cast the manifest structure into stone and metal.  Let's get it right before we do that.